Add ImGui-based developer mode to Launcher (and remove Viewer)

tgjones opened the pull request.

Qibbi commented on src/OpenSage.Game/Diagnostics/MainView.cs:
```diff
@@ -0,0 +1,165 @@
+        private readonly DiagnosticViewContext _context;
+        private readonly List<DiagnosticView> _views;
+
+        private bool _isVSyncEnabled;
```
Ideally this should be an option which is read/saved to a settings file. Also a caching thing, if the graphics device's property is used it's 4 indirections.

tgjones replied:
A settings file is a nice idea. (I notice that imgui already makes one for itself, which saves the window positions.)

I checked in Veldrid, and the property getter for `SyncToVertialBlank` is just a simple getter (albeit hidden behind another getter), so it's not going to be a perf issue to get that every frame.

tgjones commented on src/OpenSage.Game/Content/MapLoader.cs:
```diff
@@ -182,6 +182,8 @@ Texture LoadTexture(string name)
 
             return new Scene3D(
```
Sounds sensible.

tgjones commented on src/OpenSage.Game/Diagnostics/RenderedView.cs:
```diff
@@ -0,0 +1,115 @@
+                _renderTarget.EnsureSize(newSize);
+                _cachedSize = currentSize;
+
+                _scene3D?.Camera.OnViewportSizeChanged();
```
No, it can't be null (although it could in an earlier iteration of the code).

tgjones commented on src/OpenSage.Game/Graphics/Rendering/RenderPipeline.cs:
```diff
@@ -433,7 +425,7 @@ void setDefaultConstantBuffer(string name, DeviceBuffer buffer)
             };
 
-            var cameraPosition = Matrix4x4Utility.Invert(context.Camera.View).Translation;
+            var cameraPosition = Matrix4x4Utility.Invert(context.Scene3D.Camera.View).Translation;
```
No, it's not.

tgjones commented:
Thanks @paavohuhtala. I've pushed a commit that addresses your feedback:

> Some maps in the jump list are displayed as MISSING: 'Map:NoTest', e.g Bridge Busters. Is it a bug or just unclear?

That is actually the name as stored in MapCache.ini, probably because those maps are never shown to the user. I've added the map filename to the Jump menu, so we always know what map it is.

> The asset list is a bit hard to read now. Perhaps the asset type should be in a separate column?

In C&C3 and beyond, the asset names are in that format anyway ("GameObject:TheGameObjectName" etc.) So I thought I'd use the same convention for pre-C&C3 games. I've increased the width of the column so at least you can read most of the asset names now without relying on the tooltips.

> Not sure if this was already present in master, but audio asset preview should reset the playback position to the start after the playback has finished. Is this possible with the current sound API?

Doesn't seem possible with the current API. @feliwir will this come later?

> Would it make sense to rename W3DContainer to Mesh in the asset list? It's an unfamiliar term for people who don't know SAGE in depth.

This is again a C&C3 thing. They use the names:

* `W3DMesh` - the mesh
* `W3DCollisionBox` - the collision box
* `W3DHierarchy` - the bone hierarchy
* `W3DContainer` - one `W3DHierarchy`, and a collection of `W3DMesh` and `W3DCollisionBox` references

So the thing that's currently in the asset list corresponds to `W3DContainer`. I do also plan to list the `W3DMesh`es separately, because that will help with unifying the code between SAGE 1.0 and "SAGE 1.5".

> This is probably impossible now, but when we refactor the content manager to track which assets are loaded for which maps we should add that information (why this asset is loaded) to the asset list.

That's a good idea. (In SAGE 1.5, I don't think there's any difference between those two things.)

Qibbi commented:
> So the thing that's currently in the asset list corresponds to `W3DContainer`. I do also plan to list the `W3DMesh`es separately, because that will help with unifying the code between SAGE 1.0 and "SAGE 1.5".
Both W3DContainer and W3DMesh can be referenced as the model of a GameObject's draw for example, so this makes sense either way.

feliwir commented:
@tgjones
Should be possible, but i want to have proper seeking support in SharpAudio first.

tgjones commented:
@Qibbi That's good to know.

tgjones commented:
Now there's no more standalone Viewer, and the Launcher has a developer mode (activated with F11) that looks like this:

![unknown](https://user-images.githubusercontent.com/102049/51064156-28abda00-15f6-11e9-9d88-0d781211809a.png)
